File: src/main/java/discretemath/structure/function/concrete/FactorialFunction.java
package discretemath.structure.function.concrete;

import java.math.BigInteger;

import discretemath.structure.sequence.CacheableSequence;
import discretemath.structure.sequence.RecurrentSequence;
import discretemath.structure.sequence.Sequence;
import discretemath.structure.set.Sets;

public final class FactorialFunction extends RecurrentSequence<BigInteger> {

	private FactorialFunction() {
		super(Sets.Z_PLUS);
	}

	public static Sequence<BigInteger> get() {
		return new CacheableSequence<>(new FactorialFunction());
	}

	@Override
	public boolean isOnto() {
		return false;
	}

	@Override
	public boolean isOneToOne() {
		return false;
	}

	@Override
	protected BigInteger getNextRecurrentValue(int n) {
		if (n == 0 || n == 1) {
			return BigInteger.ONE;
		}

		return getNextRecurrentValue(n - 1).multiply(BigInteger.valueOf(n));
	}
}

File: src/main/scala/discretemath/structure/function/concrete/FactorialFunction.scala
package discretemath.structure.function.concrete

import discretemath.structure.sequence.{CacheableSequence, RecurrentSequence}
import discretemath.structure.set.Sets

import scala.math.BigInt

final class FactorialFunction private() extends RecurrentSequence[BigInt](Sets.Z_PLUS) {
  override def isOnto = false

  override def isOneToOne = false

  override protected def getNextRecurrentValue(n: BigInt): BigInt = {
    if ((n == 0) || (n == 1)) return 1

    getNextRecurrentValue(n - 1) * n
  }
}

object FactorialFunction {
  def apply = new CacheableSequence[BigInt](new FactorialFunction)
}
